keps/sig-node/20190517-image-decryption.md

```diff
@@ -0,0 +1,314 @@
+
+## Proposal
+
+The initial design includes:
+
+- `ImageDecryptSecret` API resource definition
+- `ImageDecryptSecret` pod field for specifying the ImageDecryptSecret the pod should be run with
+- Kubelet implementation for fetching & interpreting the ImageDecryptSecret
+- CRI API & implementation for passing along the ImageDecryptSecret
+
+### API
+
+`ImageDecryptSecret` is a new cluster-scoped Secret
+
+
+The private key is selected by the pod by specifying the ImageDecryptSecret in the PodSpec. Once the pod is
+scheduled, the ImageDecryptSecret cannot be changed.
+
+_(This is a simplified declaration, syntactic details will be covered in the API PR review)_
+
+```go
+// Container represents a single container that is expected to be run on the host.
+type Container struct {
+  <snip>
+  // ImageDecryptSecrets is an optional list of references to secrets in the same namespace to use for decrypting any of the images used by this PodSpec.
+  // If specified, these secrets will be passed to individual puller implementations for them to use.
+  // +optional
+  ImageDecryptSecrets []LocalObjectReference
+  </snip>
+}
+```
+[Reference Implementation](https://github.com/harche/kubernetes/blob/8a0ef9898a55110d7e41f8e1a846c66cfc2fa691/pkg/apis/core/types.go#L1915) of the `Container` struct
+
+```go
+const (
+  <snip>
+  // SecretTypeDecryptKeys defines the type for the decrypt secrets
+  SecretTypeDecryptKey SecretType = "kubernetes.io/decryptionkey"
+
+  // ImageDecryptionKeys represent the key required to access secret data
+  ImageDecryptionKey = ".imagedecryptionkey"
+  </snip>
+)
+```
+[Reference Implementation](https://github.com/harche/kubernetes/blob/8a0ef9898a55110d7e41f8e1a846c66cfc2fa691/staging/src/k8s.io/api/core/v1/types.go#L4986) for the `constant` that hold the actual private key required for decrypting the image
+
+
+In order to support ImageDecryptSecret using service account, we need to define a struct that holds the ImageDecryptSecret and the corresponding image(s) that can be decrypted.
+
+
+```go
+// ImageDecryptServiceSecret holds the name of the decrypt secret and the corresponding list of images that can be decrypted using that secret.
+type ImageDecryptServiceSecret struct {
+	// Required.
+	Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
+	// Required.
+	Images []string `json:"images,omitempty" protobuf:"bytes,8,rep,name=images"`
+}
+```
+See the [reference implementation](https://github.com/harche/kubernetes/blob/8a0ef9898a55110d7e41f8e1a846c66cfc2fa691/staging/src/k8s.io/api/core/v1/types.go#L3627) for `ImageDecryptServiceSecret`
+
+```go
+type ServiceAccount struct {
+  <snip>
+  // ImageDecryptSecrets is a list of references to secrets in the same namespace to use for decrypting any encrypted images
+  // in pods that reference this ServiceAccount. ImageDecryptSecrets are distinct from Secrets because Secrets
+  // can be mounted in the pod, but ImageDecryptSecrets are only accessed by the kubelet.
```

Comment:
Can ImageDecryptSecrets be mounted in a pod?  If not, how is this enforced?

Reply:
Since `ImageDecryptSecrets` are modelled after `ImagePullSecrets`, they can be mounted in a pod. But considering the sensitive nature of the data that both of these secrets carry maybe it's a good idea to prevent them from getting mounted in a pod. 

We are thinking about bringing generic enough authorization check while starting the containers, also suggested by Derek below in his [comment](https://github.com/kubernetes/enhancements/pull/1066#discussion_r301293131), for both `ImageDecryptSecrets` as well as `ImagePullSecrets`. I think we should extend that to include prevention both of these secrets from getting mounted in a pod. 

But I am not sure if this particular KEP should handle it, because the solution has to be generic enough for any secret that's only consumable by the runtime and not the pod itself.

Reply:
That makes sense - I agree with the idea of having some form of protection that can key off the secret type, and I agree that this could be done separately.  I think what threw me off is that in the comment you seem to be saying that the enforcement is in-scope, but I think you're actually saying "ImageDecryptSecrets (like ImagePullSecrets) are expected to be accessed by the kubelet, not by pods"